packages/dev/core/src/Meshes/csg2.ts

```diff
@@ -0,0 +1,513 @@
+            const extents = output.getBoundingInfo().boundingSphere.center;
+            output.position.set(-extents.x, -extents.y, -extents.z);
+            output.bakeCurrentTransformIntoVertices();
+        }
+
+        // Submeshes
+        let id = manifoldMesh.runOriginalID[0];
+        let start = manifoldMesh.runIndex[0];
+        let materialIndex = 0;
+        const materials: Material[] = [];
+        scene = output.getScene();
+        for (let run = 0; run < manifoldMesh.numRun; ++run) {
+            const nextID = manifoldMesh.runOriginalID[run + 1];
+            if (nextID !== id) {
+                const end = manifoldMesh.runIndex[run + 1];
+                new SubMesh(materialIndex, 0, vertexCount, start, end - start, output);
+                materials.push(scene.getMaterialByUniqueID(id - FirstID) || scene.defaultMaterial);
+                id = nextID;
+                start = end;
+                materialIndex++;
+            }
+        }
+
+        if (materials.length > 1) {
+            const multiMaterial = new MultiMaterial(name, scene);
+            multiMaterial.subMaterials = materials;
+            output.material = multiMaterial;
+        } else {
+            output.material = materials[0];
+        }
+
+        return output;
+    }
+
+    /**
+     * Dispose the CSG resources
+     */
+    public dispose() {
+        if (this._manifold) {
+            this._manifold.delete();
+            this._manifold = null;
+        }
+    }
+
+    private static _ProcessData(
+        vertexCount: number,
+        triVerts: Uint32Array,
+        structure: IManifoldVertexComponent[],
+        numProp: number,
+        runIndex?: Uint32Array,
+        runOriginalID?: Uint32Array
+    ) {
+        const vertProperties = new Float32Array(vertexCount * structure.reduce((acc, cur) => acc + cur.stride, 0));
+
+        for (let i = 0; i < vertexCount; i++) {
+            let offset = 0;
+            for (let idx = 0; idx < structure.length; idx++) {
+                const component = structure[idx];
+
+                for (let strideIndex = 0; strideIndex < component.stride; strideIndex++) {
+                    vertProperties[i * numProp + offset + strideIndex] = component.data![i * component.stride + strideIndex];
+                }
+                offset += component.stride;
+            }
+        }
+
+        const manifoldMesh = new ManifoldMesh({ numProp: numProp, vertProperties, triVerts, runIndex, runOriginalID });
+        manifoldMesh.merge();
+
+        try {
+            return new CSG2(new Manifold(manifoldMesh), numProp, structure);
+        } catch (e) {
+            throw new Error("Error while creating the CSG: " + e.message);
+        }
+    }
+
+    /**
+     * Create a new Constructive Solid Geometry from a vertexData
+     * @param vertexData defines the vertexData to use to create the CSG
+     * @returns a new CSG2 class
+     */
+    public static FromVertexData(vertexData: VertexData): CSG2 {

```

Comment:
Cound technically be outside of this class (unless it is always required and make no sense to not include it)

Reply:
Indeed, always required (I thought about it :) but in this case it is strongly linked so no reason to make it more difficult)